File: x_gbif_download.jl
using SimpleSDMLayers
using GBIF
using DataFramesMeta
using StatsBase
import CSV

# Download and parse the metaweb
mw_path = joinpath("data", "input", "canadian_thresholded.csv")
mw_output = DataFrame(CSV.File(mw_path; stringtype=String))

# Turn the metaweb into a network
sp = GBIF.taxon.(unique(vcat(mw_output.from, mw_output.to)))

# Find potential duplicates
spdf = DataFrame(sp)
select!(spdf, Not([:kingdom, :phylum, :class, :order, :family, :genus]))
id = [sp.second for sp in spdf.species]
unique(id)
spdf.duplicate = [sum(sp .== spdf.species) > 1 for sp in spdf.species]
@chain begin
    spdf
    @rsubset(:duplicate == true)
    show(_, allrows=true)
end

# What happens when we have duplicates?
occurrences(GBIF.taxon("Canis rufus", strict=true), "hasCoordinate" => true)
occurrences(GBIF.taxon("Canis lupus", strict=true), "hasCoordinate" => true)
# Same occurrences

## Evaluate number of observations
# How does this work
sp1 = occurrences(sp[1], "hasCoordinate" => true)
length(sp1) # downloaded obs
size(sp1) # total obs

# Get occurrences for all species
occ = Vector{GBIFRecords}(undef, length(sp))
for i in eachindex(sp)
    occ[i] = occurrences(
        sp[i],
        "hasCoordinate" => true,
        "limit" => 1,
        "decimalLatitude"=>"10,90",
        "decimalLongitude" => "-175,-45"
    )
end
occ
occdf = select(spdf, :name)
occdf.nocc = [size(o) for o in occ]
sort(occdf, :nocc, rev=true) |> x -> show(x, allrows=true)
# Queries seem limited at 100_000 obs
# Some species have less than 100 obs

# Check the current file size
occdf.ncurrent = zeros(Int64, nrow(occdf))
for r in eachrow(occdf)
    _spname = replace(r.name, " " => "_")
    _df = CSV.read(joinpath("data", "occurrences", _spname*".csv"), DataFrame)
    r.ncurrent = nrow(_df)
end
sort(occdf, :nocc, rev=true) |> x -> show(x, allrows=true) # makes no sense...
# Ah, probably because of the coordinates!
# Still doesn't make much sense
# Can't only be because of recent increase in number of observations?
# Maybe taxonomic backbone update?

# How many more observations do we have
sum(occdf.nocc) # 1,780,025
sum(occdf.ncurrent) # 159,471

## Extract the keys
spdf.key = [sp.second for sp in spdf.species]
spdf
spkey = @chain begin
    leftjoin(occdf, spdf, on = :name)
    select(:name, :nocc, :ncurrent, :key)
    sort(:nocc, rev=true)
end
[spkey.key] # copy paste this in curl command

## Read the downloaded data

# Read a smaller dataset to precompile
df = DataFrame(CSV.File(joinpath("data/input/canadian_thresholded.csv")))

# Attempt to read the complete dataset (should all fail)
occ_path = joinpath("data", "occurrences", "subset", "all_occurrences.csv")
df = DataFrame(CSV.File(occ_path))
df = DataFrame(CSV.File(occ_path; stringtype=String))
df = DataFrame(CSV.File(occ_path, delim="\t", types=Dict(:gbifID => String)))
df = DataFrame(CSV.File(occ_path, delim="\t", types=Dict(:gbifID => String), quotechar='"'))
df = DataFrame(CSV.File(occ_path, select=[:species, :decimalLatitude, :decimalLongitude]))

# Attempt with fewer rows
df = DataFrame(CSV.File(occ_path, limit=100))
df = DataFrame(CSV.File(occ_path, limit=100, delim="\t"))
df = DataFrame(CSV.File(occ_path, limit=100, delim="\t", types=Dict(:gbifID => String)))
df = DataFrame(CSV.File(occ_path, limit=1_000)) # not reading the right number of rows...
df = DataFrame(CSV.File(occ_path, limit=11_800)) # when it stops working

# Attempt with dataset exported with R
df = DataFrame(CSV.File(joinpath("data", "occurrences", "subset", "all_occurrences2.csv"), delim="\t"))
# Not 100_000 rows??
# Same with R though...

# Attempt with problem dataset
df = DataFrame(CSV.File(joinpath("data", "occurrences", "subset", "all_occurrences_problems.csv"), delim="\t"))

File: x_gbif_download.R
library(tidyverse)

# Load complete dataset
occ <- read_tsv("./data/occurrences/subset/all_occurrences.csv") # works now??

# Create a smaller data set
write_csv(head(occ), "data/occurrences/subset/all_occurrences_mini.csv")

# Try to reexport
write_tsv(head(occ, 100000), "data/occurrences/subset/all_occurrences2.csv")

# Check exported dataset
occ2 <- read_tsv("./data/occurrences/subset/all_occurrences2.csv") # wrong number of rows??
head(occ, 100000) %>% nrow()

# Investigate problems
problem_rows <- problems(occ2) %>% pull(row)
problem_rows_buffered <- sort(c(problem_rows-1, problem_rows, problem_rows+1))
problem_occ <- slice(occ, problem_rows_buffered)
problem_occ
write_tsv(problem_occ, "./data/occurrences/subset/all_occurrences_problems.csv")

# Re-read problems?
occ3 <- read_tsv("./data/occurrences/subset/all_occurrences_problems.csv")
all(occ3 == problem_occ, na.rm = TRUE)
occ3 == problem_occ %>% View()
occ3[which(occ3 != problem_occ)

## Temporary workaround
# Keep only basic columns, separate by species, export to separate files

# First use a smaller set
# occ_mini <- head(occ, 1000)

# Now select only the same columns as in your initial analyses
# occ_select <- occ_mini |>
occ_select <- occ |>
  select(species, decimalLatitude, decimalLongitude) |>
  rename(name = species,
         latitude = decimalLatitude,
         longitude = decimalLongitude)

# Separate by species
# occ_select %>%
#   group_by(name) %>%
#   # summarize(test = length(name)) %>%
#   group_split()

# Get species names
sp_list <- unique(occ_select$name)

# Export as separate files
for (sp in sp_list) {
  sp_name <- str_replace(sp, " ", "_")
  sp_path <- file.path("data", "occurrences", paste0(sp_name, ".csv"))
  occ_select |>
    filter(name == sp) |>
    write_tsv(sp_path)
}
# 67 MB
# 1,579,706 obs
